File: Second Year/Semester II/Web Programming/lab8_aspnet/lab8_aspnet/Services/RoomService.cs
using lab8_aspnet.DTOs;
using lab8_aspnet.Models;
using lab8_aspnet.Repositories.Interfaces;
using lab8_aspnet.Services.Interfaces;

namespace lab8_aspnet.Services
{
    public class RoomService : IRoomService
    {
        private readonly IRoomRepository _roomRepository;

        public RoomService(IRoomRepository roomRepository)
        {
            _roomRepository = roomRepository;
        }

        private static RoomDto ToDto(Room room) => new RoomDto
        {
            Id = room.Id,
            HotelName = room.Hotel!.Name,
            RoomNumber = room.RoomNumber,
            Category = room.Category,
            Type = room.Type,
            Price = room.Price
        };

        public async Task<List<RoomDto>> GetRoomsAsync(
            string? category, string? type, decimal? maxPrice, string? hotelName) =>
            (await _roomRepository.GetRoomsAsync(category, type, maxPrice, hotelName)).Select(ToDto).ToList();

        public async Task<RoomDto> InsertRoomAsync(RoomCreateRequest request)
        {
            var room = new Room
            {
                HotelId = request.HotelId,
                RoomNumber = request.RoomNumber,
                Category = request.Category,
                Type = request.Type,
                Price = request.Price
            };

            var createdRoom = await _roomRepository.InsertRoomAsync(room);
            return ToDto(createdRoom);
        }

        public async Task<RoomDto?> UpdateRoomAsync(int id, RoomUpdateRequest request)
        {
            var updatedRoom = await _roomRepository.UpdateRoomAsync(id, request.RoomNumber, request.Category, request.Type, request.Price);
            return updatedRoom == null ? null : ToDto(updatedRoom);
        }

        public async Task<bool> DeleteRoomAsync(int id) =>
            await _roomRepository.DeleteRoomAsync(id);
    }
}

File: Second Year/Semester II/Web Programming/lab9_node/src/services/roomService.js
const roomRepository = require('../repositories/roomRepository');

function toDto(room) {
    return {
        id: room.id,
        hotel_name: room.hotelName,
        room_number: room.roomNumber,
        category: room.category,
        type: room.type,
        price: room.price
    };
}

async function getRooms(category, type, maxPrice, hotelName) {
    const rooms = await roomRepository.getRooms(category, type, maxPrice, hotelName);
    return rooms.map(toDto);
}

async function insertRoom(hotelId, roomNumber, category, type, price) {
    const idCreatedRoom = await roomRepository.insertRoom(hotelId, roomNumber, category, type, price);
    const createdRoom = await roomRepository.getRoomById(idCreatedRoom);
    return toDto(createdRoom);
}

async function updateRoom(id, roomNumber, category, type, price) {
    const updatedRoom = await roomRepository.updateRoom(id, roomNumber, category, type, price);
    if (!updatedRoom) {
        return null;
    }
    const room = await roomRepository.getRoomById(id);
    return toDto(room);
}

async function removeRoom(id) {
    return await roomRepository.removeRoom(id);
}

module.exports = {
    getRooms,
    insertRoom,
    updateRoom,
    removeRoom
};